composer/utils/object_store/sftp_object_store.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Utility for uploading to and downloading from cloud object stores."""
+import os
+import pathlib
+import urllib.parse
+import uuid
+from typing import Callable, Optional, Union
+
+from paramiko import SSHException
+
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore, ObjectStoreTransientError
+
+__all__ = ['SFTPObjectStore']
+
+
+class SFTPObjectStore(ObjectStore):
+    """Utility for uploading to and downloading to a server via SFTP.
+
+    Args:
+        host (str): The server to connect to. Also accepts a string in the form 'username@host'
+        port (int, optional): The server port to connect to. Defaults to 22
+        username (str, optional): The username to authenticate. Raises an exception if username is not
+            passed in and also cannot be read from the host argument.
+        key_file_path (str, optional): The filename of the private key.
+        cwd (str, optional): The directory to navigate to upon creating the SSH connection. If not present
+            it will be created.
+    """
+
+    def __init__(self,
+                 host: str,
+                 port: int = 22,
+                 username: Optional[str] = None,
+                 key_file_path: Optional[str] = None,
+                 cwd: Optional[str] = None):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.key_file_path = key_file_path
+        self.cwd = cwd
+
+        if host.startswith('sftp://'):
```

Comment:
nit: Can we always attempt to `urllib.parse.urlparse(host)`, and then check to see if `scheme` is empty (in which case it's just the host), or `sftp` (in which case it's the URI)